### PR TITLE
Backend tests

### PR DESCRIPTION
**Pull request by jacobrs**
Didn't manage to mock the append on DB models, will need to work on that but probably in another PR. This should increase the test coverage by quite a bit hopefully.

### CODE REVIEW

File: server/src/rest_service.py [hunk 1]
```diff
@@ -63,7 +63,7 @@ def add_team(db_session, data):
         return True, "Successfully added Team {}".format(data["name"])
     except Exception as e:
         db_session.rollback()
-        return False, "Could not add Team {}; already exists".format(data["name"])
+        return False, "Could not add Team {}; already exists".format(e)
```
Comment (jacobrs) [inline server/src/rest_service.py]: I thought that I should revert this change but thinking about it now I think I should keep it. The e gives better contextual feedback since its not always an already exists error.

File: server/src/mocks.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+import schema
+
```
Comment (lancelafontaine) [inline server/src/mocks.py]: Nice, like this much better than redefining stubs for every test.

File: server/src/rest_service.py (hunk 1 above)
Comment (lancelafontaine) [reply server/src/rest_service.py]: Sure, but then we should remove the part of the string that says `already exists` and only display the error message.

File: server/src/run.py [hunk 3]
```diff
@@ -2,6 +2,7 @@
 from sanic_cors import CORS, cross_origin
 from config import config
 import controller
+import schema
```
Comment (lancelafontaine) [inline server/src/run.py]: Is this necessary? No other changes were made to this file.

**Review by lancelafontaine — APPROVED** [review]
Left a few small comments but overall looks good to me 👍

File: server/src/run.py (hunk 3 above)
Comment (jacobrs) [reply server/src/run.py]: oops forgot to remove that, was messing around with some stuff. Good catch!

File: server/src/rest_service.py (hunk 1 above)
Comment (jacobrs) [reply server/src/rest_service.py]: true that!

File: server/src/controller_test.py [hunk 4]
```diff
@@ -26,6 +29,24 @@ def json(self):
     data = json.loads(controller.get_slack_users().body)
     assert data == {'message': 'Slack API error', 'ok': False}
 
+def test_get_app_users():
+    mockUser = mocks.MockUser("testuser", "testpassword", "tester", 0, "sudoers")
+    mockRequest = {"username": "testuser", "team": "sudoers"}
+
+    result = controller.get_app_users(mockRequest, mocks.MockFilterProvider(mockUser))
+    data = json.loads(result.body)
+    assert data['ok'] == True
```
Comment (tamarindd) [inline server/src/controller_test.py]: you should use 'is' here instead of '==' (just because it's more pythonic, doesn't affect functionality)

File: server/src/controller_test.py [hunk 5]
```diff
@@ -26,6 +29,24 @@ def json(self):
...
+    assert data['data'][0]['username'] == mockRequest["username"]
+
+def test_get_teams():
+    mockTeam = mocks.MockTeam("team1", [mocks.MockUser("testuser")])
+    mockRequest = {"name": "team1"}
+
+    result = controller.get_teams(mockRequest, mocks.MockFilterProvider(mockTeam))
+    data = json.loads(result.body)
+    assert data['ok'] == True
+    assert data['data'][0] != None
```
Comment (tamarindd) [inline server/src/controller_test.py]: is not None instead of != None

File: server/src/controller_test.py [hunk 6]
```diff
@@ -89,6 +110,81 @@ def json(self):
...
+                        'profile': {
+                            'real_name_normalized': 'I should throw',
+                            'image_192': 'https://fakeurl.coolbeans'
+                        }
+                    }
+                ]
+            }
+    monkeypatch.setattr(requests, 'get', SlackUsersMock)
+    response = controller.get_slack_users()
+    assert type(response) is HTTPResponse
```
Comment (tamarindd) [inline server/src/controller_test.py]: isinstance(response, HTTPResponse) is better in general because it will return true if the type is a child of the one you're checking against

File: server/src/mocks.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
+class OneResult:
+    def __init__(self, result = None):
+        self.result = result
```
Comment (tamarindd) [inline server/src/mocks.py]: you should really put empty lines between methods (just to improve readability), but don't bother if it's too much trouble

File: server/src/mocks.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+class MockTeam:
+    def __init__(self, name = None, users = None):
+        self.name = name
+        self.members = users
+
+class MockAdderDatabase:
+    def add(self, user):
+        if user.username == "fail":
+            raise Exception("db simulated failure")
+        return
```
Comment (tamarindd) [inline server/src/mocks.py]: this is pointless here

File: server/src/mocks.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    def commit(self):
+        return
+
+class FaultyCommitDatabase:
+    def query(self, filter):
+        return ShallowFilter(MockUser("anything"))
+    def rollback(self):
+        return
+    def commit(self):
+        raise Exception("db simulated failure")
```
Comment (tamarindd) [inline server/src/mocks.py]: maybe raise an SQLAlchemy exception instead? you can never be too specific.
http://docs.sqlalchemy.org/en/latest/core/exceptions.html#sqlalchemy.exc.SQLAlchemyError

File: server/src/schema_test.py [hunk 10]
```diff
@@ -0,0 +1,7 @@
+import schema
+import pytest
+
+def test_repr():
+	assert repr(schema.Team()) != ""
```
Comment (tamarindd) [inline server/src/schema_test.py]: if repr(..) is an empty string it will evaluate to False

File: server/src/mocks.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+        return [self.dataToReturn]
+
+class MockFilterProvider:
+    def __init__(self, data = None):
+        self.data = data
+    def query(self, any):
+        return ShallowFilter(self.data)
+
+class MockUser:
+    def __init__(self, username = None, password = None, name = None, score = None, team = None):
```
Comment (tamarindd) [inline server/src/mocks.py]: the convention is not to surround '=' with spaces for default parameters

File: server/src/controller_test.py (hunk 4 above)
Comment (tamarindd) [reply server/src/controller_test.py]: actually, there's really no need to compare any of these values to booleans or None. you can just do 'assert data['ok']' or 'assert not data['ok']'

**Review by tamarindd — APPROVED** [review]
pretty much all my comments are about python standards, so not a big deal. thanks for writing my tests :P

**Comment by jacobrs** [comment]
Wow, sooo many python standards, but some comments are actually valid. Not a fan of adding so much whitespace between functions since the forced indenting makes it obvious.

File: server/src/controller_test.py (hunk 4 above)
Comment (jacobrs) [reply server/src/controller_test.py]: thats true!
